src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+# Proposal: QIR Adaptor Tool Specification
+
+This document discuss a tool that transforms a QIR into a restricted version of the QIR (known as a profile).
+We aim to make a specification for a generic tool that allows the user to
+
+1. Create or use an existing profile without the need of writing code.
+2. Validate that a QIR is compliant with the specific profile.
+3. Generate a profile compliant QIR from a generic unconstrained QIR (if possible).
+
+This document sets out to motivate and demonstrate feasibility of building such a tool.
+
+## Motivation
+
+It is anticipated that most usages of the QIR specificaiton will need to only use a subset of it. These subsets
+may further be subject to constraints such as how one allocate or aquire a qubit handle. We refer subset with
+constraints as a profile. For instance, it is likely that early versions of quantum hardware will have a limited
+set of classical instructions available. With this in mind, the vendor or user of said hardware would define a profile
+that only contains a specified subset. One example of such a profile, is the [base profile](https://github.com/microsoft/qsharp-language/blob/ageller/profile/Specifications/QIR/Base-Profile.md)
+which only allows function calls and branching, but no arithmetic, classical memory or classical registers.
```

Comment:
I find this sentence a bit confusing, since the profile does allow to store measurement results as classical bits such that it is possible to branch based on them. Additionally, the profile also allows for integer and double literals (constants). Saying there is no classical memory is hence a bit odd to me.

Reply:
I few points here:
1. It does not allow you to store or even look at measurement results as far as I can see. There is exactly one function to perform measurements which allows you to flush the result(with an assigned id) but no means of retrieving the result. See the function signature of `__quantum__qis__mz__body`. You would therefore not be able to branch on any measurement result. I would argue that there is a classical sink.
2. Constants are neither registers nor memory. Yes, literals are allowed
3. Given the classical instruction set `call`, `br` and `ret`, but no means of retrieving, storing or in any way manipulating data. Consequently, only sensible function you can have will be `Unit` functions